TCO calculator rebuild

bethcollins92 opened the pull request:
## Done

- Remove logic needed for old tco-calculator
- Add react component for new calculator, all calculations are based on this [spreadsheet](https://docs.google.com/spreadsheets/d/1KrR3s21K2uyjYtwHHnPHiN91byK5SXKxU-NXdCA8MbY/edit#gid=1544469160)
- Update secondary nav and path to `openstack/pricing-calculator`, and add a redirect from `/openstack/tco-calculator`
- Build new page including new calculator
- Integrate enzyme testing framework and add tests
- Add new contact us form for the estimates - for MVP there is no option to add another instance


## QA

- View the page at: https://ubuntu-com-9974.demos.haus/openstack/pricing-calculator
- First check the copy and design based on this [copy doc](https://docs.google.com/document/d/1n3ExZwFVlw5LyAp7O3CgPthVgx8yRX5-Wg_gBQLf73c/edit#) and this [design](https://app.zeplin.io/project/60d33b8cbfc46b12b0b4d3cb/screen/60d33babdf4648121efe1e13)
- Go to the [spreadsheet](https://docs.google.com/spreadsheets/d/1KrR3s21K2uyjYtwHHnPHiN91byK5SXKxU-NXdCA8MbY/edit#gid=1544469160), and change the inputs on the `Calculator` tab. (e.g. Number of instances, vCPUs etc)
- Go back to https://ubuntu-com-9974.demos.haus/openstack/pricing-calculator and change the values on the inputs on the calculator to be exactly the same as the spreadsheet. Check the outputs are the same. 

- NOTE: `Get the free report` and `Learn more about TCO calculators` links to a copy doc of an engage page which doesn't exist yet. 

## Issue / Card

Fixes [#4197](https://github.com/canonical-web-and-design/web-squad/issues/4197)

## Screenshots

![Screenshot 2021-06-30 at 11 12 44](https://user-images.githubusercontent.com/58959073/123943498-1ffd1900-d994-11eb-8444-645d059c377f.png)

webteam-app commented:
Demo starting at https://ubuntu-com-9974.demos.haus

anasereijo commented:
Hi, @bethcollins92 this is almost there! Just some amendments:

- [ ] make the title separate from the h4 and card as per design

<img width="1117" alt="Screenshot 2021-06-30 at 11 38 34" src="https://user-images.githubusercontent.com/36884067/123947633-7f5d2800-d998-11eb-8949-9222ead8de35.png">

- [ ] check headings (font sizes) within the calculator - I checked the copydoc and they match the design

- [ ] number of instances: can we have the ranges as per design? see below

<img width="335" alt="Screenshot 2021-06-30 at 11 45 47" src="https://user-images.githubusercontent.com/36884067/123947869-c3e8c380-d998-11eb-82b2-339cb0a08ff0.png">

- [ ] have an empty col between the 2 cols. I believe there is a class for that

<img width="1089" alt="Screenshot 2021-06-30 at 11 47 16" src="https://user-images.githubusercontent.com/36884067/123948247-1fb34c80-d999-11eb-995a-380064f0ea32.png">

<img width="1065" alt="Screenshot 2021-06-30 at 11 48 06" src="https://user-images.githubusercontent.com/36884067/123948264-2346d380-d999-11eb-912c-cce1f06c823b.png">

- [ ] I think we need errors messages if the user adds values that are not in the range of each item?

<img width="1088" alt="Screenshot 2021-06-30 at 11 50 43" src="https://user-images.githubusercontent.com/36884067/123948678-97817700-d999-11eb-8793-0b084f8ab948.png">

- [ ] fix hr spacing: should be 32px

<img width="1134" alt="Screenshot 2021-06-30 at 11 52 30" src="https://user-images.githubusercontent.com/36884067/123948940-e0393000-d999-11eb-8ff3-93c9f1c6577e.png">

- [ ] Email estimates button is missing

<img width="586" alt="Screenshot 2021-06-30 at 11 54 39" src="https://user-images.githubusercontent.com/36884067/123949027-fe069500-d999-11eb-8bcb-f1b7a6c4e3de.png">


- [ ] as we have a max value can we reduce the fields a bit? 

<img width="467" alt="Screenshot 2021-06-30 at 11 55 59" src="https://user-images.githubusercontent.com/36884067/123949346-5d64a500-d99a-11eb-8e5e-4563e861397f.png">

<img width="557" alt="Screenshot 2021-06-30 at 11 57 46" src="https://user-images.githubusercontent.com/36884067/123949383-68b7d080-d99a-11eb-9843-77bf7c8eaab4.png">

ClementChaumel commented:
Consider adding a max value for the inputs :sweat_smile: 

![maxcalculator](https://user-images.githubusercontent.com/11927929/123958153-01574c00-d9ad-11eb-942a-282086c16e90.gif)

ClementChaumel commented on static/js/src/openstack/react/components/CostCalculations.jsx:
```diff
@@ -0,0 +1,293 @@
+  //RAM
+  const amountOfRamInCloud = formState.instances * formState.ram;
+
+  const amountOfRamInFullyUtilizedCloud = Math.ceil(
```
You use Math.ceil for each intermediate value. Is it defined this way in the spreadsheet? Otherwise you could only use Math.ceil on the final computed value to minimise rounding errors

bethcollins92 replied:
It is defined like this on the spreadsheet, but I think it makes sense to leave it on the final computed value like you say will reduce rounding errors

ClementChaumel commented on static/js/src/openstack/react/components/CostCalculations.jsx:
```diff
@@ -0,0 +1,293 @@
+
+import TCO_VARIABLES from "../utils/variables";
+
+const CostCalculations = (formState) => {
```
```suggestion
const CostCalculations = ({ instances, vcpus, emepheralStorage, ram, persistentStorage, supportLevel}) => {
```
In here I think you could access formstate attribute directly (and I think eslint would require you to add proptypes for each of the parameters)

ClementChaumel commented on static/js/src/openstack/react/utils/variables.js:
```diff
@@ -0,0 +1,56 @@
+const TCO_VARIABLES = {
```
Since they don't change, shouldn't it be TCO_CONSTANTS?

bethcollins92 replied:
makes sense 😅

ClementChaumel commented:
Also, 0 instances cost me infinity$ !! :money_with_wings:  
![image](https://user-images.githubusercontent.com/11927929/123960384-86436500-d9af-11eb-8ede-86e96e98c1c2.png)

ClementChaumel commented:
For the tests, you should run `test-js` instead of `test`. If I try it with dotrun it ""works"" (it doesn't but I get a relevant error message)
![image](https://user-images.githubusercontent.com/11927929/123961204-6496ad80-d9b0-11eb-93c9-7f1e6238f38d.png)

bethcollins92 commented:
@anasereijo I've made all changes possible without having to write lots of custom css, as we are pushed for time with this and there are still tests to write. @petesfrench is now taking over so please contact him with any questions. Thank you!

bethcollins92 commented:
@anasereijo Could you take another look at this please? I've addressed your comments, however: 

- I can't add numbers to the side of the slider without a bunch more code as the vanilla component doesn't have an option for this. 
- If I make the size of the inputs smaller the error messages also stay to that width so it looks awful. If this is a big issue and they need to be smaller I can look into it but it will take a lot more work.

anasereijo commented:
- [ ] headings aren't matching the design
how it looks like:
<img width="1100" alt="Screenshot 2021-07-14 at 09 22 22" src="https://user-images.githubusercontent.com/36884067/125589082-be249c3c-9288-4977-9a81-d6d75f5ef9e9.png">

how it should look like:
<img width="1085" alt="Screenshot 2021-07-14 at 09 22 16" src="https://user-images.githubusercontent.com/36884067/125589116-1502a114-e7a4-47ee-ae0c-981c543c946b.png">

- [ ] add a bit more padding here
<img width="557" alt="Screenshot 2021-07-14 at 09 24 22" src="https://user-images.githubusercontent.com/36884067/125589332-5fad5d46-b757-4a50-bc65-5ade8363ec08.png">

SirSamTumless commented:
@tytus-kurek Quick heads up - to get this out this iteration we are not adding the "add another instance" in the contact form.

anthonydillon commented on static/js/src/openstack/react/components/CostCalculatorForm.jsx:
```diff
@@ -0,0 +1,223 @@
+                <Icon name="help" className="tco-tooltip" />
+              </Tooltip>
+            </h3>
+            <Row>
+              <Col size="3">
+                <Row>
+                  <Col size="3">
```
This is quite confusing and it's abusing the wrapping natural of a misconfigured set of cols in a grid. I would suggest, moving all these elements to the top-level `Col` and adding max-width style to the inputs.

anthonydillon commented on static/js/src/openstack/react/components/CostCalculatorForm.jsx:
```diff
@@ -0,0 +1,223 @@
+              </Col>
+            </Row>
+          </Col>
+          <Col size="6">
```
If you do this:
```
<Col size="5" emptyLarge="8">
```
Then you wont need the inner row and col.

anthonydillon commented on static/sass/_pattern_buttons.scss:
```diff
@@ -5,7 +5,7 @@
 
   .p-button--tag {
-    @extend %p-button-light;
+    @extend %p-button-light !optional;
```
What does optional do here?

bethcollins92 replied:
`./run` was failing because of this line and it suggested adding `!optional` to this variable

bethcollins92 replied:
(just removed it as Scott pointed out it was a new variable I just needed to rebase)

anthonydillon commented on templates/openstack/pricing-calculator.html:
```diff
@@ -0,0 +1,108 @@
+  </div>
+
+  <div class="u-fixed-width">
+    <p><small>NOTE: The private cloud pricing calculator is for informational purposes only. Exact prices may vary depending on your detailed requirements, cloud configuration and local cost conditions. The hourly cost per instance contains all CapEx and OpEx costs over a three-year period, including reference hardware, typical hosting charges, average operations staff salary costs, cloud deliver ...
```
@anasereijo would this be better as a [information notification](https://vanillaframework.io/docs/patterns/notification#information)?

For example: 
![image](https://user-images.githubusercontent.com/1413534/125773374-4e6379a0-ef94-4bb6-8dfc-1d78ac626689.png)

anasereijo replied:
excellent idea

anthonydillon commented on templates/shared/forms/interactive/openstack-cost-calculator.html:
```diff
@@ -0,0 +1,242 @@
+        <div class="col-4 u-sv3">
+          <div class="js-formfield">
+            <h3 class="p-heading--five">vCPUs</h3>
+            <input type="number" id="instance-type-vcpus" min="1" max="116" value="2" aria-label="Instance type: vCPUs">
```
Oh I thought this value would come from the selection on the previous page.

bethcollins92 replied:
I'm not sure how to grab the variables from the react component to inject them into the contact form, do you know how I could do this?

anthonydillon replied:
I don't think you can without ejecting from React.

bethcollins92 replied:
I'm still trying to figure out how to do this but all the other comments have been addressed and changes pushed. Do you think we should make another issue with this and the ability to add another instance together? Or I can continue working on it next week

anthonydillon replied:
This is an architectural issue. Maybe follow up with an issue to address this. I believe you can set the state to window variables with pulls the values when opening the modal using vanilla JS.

bethcollins92 replied:
thanks, managed to figure it out in the end! and did pretty much this ^

anthonydillon commented on static/js/src/openstack/react/components/CostCalculations.test.jsx:
```diff
@@ -0,0 +1,148 @@
+      .find("#persistent-storage")
+      .simulate("change", { target: { value: 90, name: "persistentStorage" } });
+
+    expect(formWrapper.find("#persistent-storage").props().value).toEqual(90);
```
These tests are checking the browser works. Maybe you should check in internal values are correct.

anthonydillon reviewed (approved):
Good work @bethcollins92 LGTM but have a few comments/suggestions in the code

anthonydillon commented:
@bethcollins92 did you forget to push your changes? I don't see any changes to the code.

bethcollins92 commented:
> @bethcollins92 did you forget to push your changes? I don't see any changes to the code.

Sorry yes, I wrote that comment and then remembered I wanted to look at the tests before I push - will let you know when it's ready again

tytus-kurek commented:
@SirSamTumless I'm fine as long as it continues being worked on during the next iteration. Also, will I get the cloud pricing report as a part of this iteration so that we could attach it from the new page?

bethcollins92 commented:
@anthonydillon Thanks for the review, I've addressed suggestions and comments

bethcollins92 commented:
Issue for adding another instance type [here](https://github.com/canonical-web-and-design/ubuntu.com/issues/10049)
